caliper@2.9.0 +rocm, ROCm 5.3.0: services/sampler/Sampler.cpp:31:10: fatal error: 'libunwind.h' file not found

### Steps to reproduce the issue

`caliper@2.9.0 +rocm` build is failing for me on `OLCF Frontier` using:
* `spack@develop` (a94d18ad087038f10c9a82ebbb1a1a1dddda51e8 from `Wed May 3 05:26:50 2023 -0700`)
* PrgEnv-gnu, GCC 11.2.0
* ROCm 5.3.0 (externally provided via `rocm/5.3.0` module, definitions included in Spack environment below)

This error is the same one blocking https://github.com/spack/spack/pull/35219

Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/11391146/spack.yaml.txt)

<details><summary>Concretization</summary>
<pre>
 -   l4up667  caliper@2.9.0%gcc@11.2.0+adiak~cuda~fortran+gotcha~ipo+libdw~libpfm+libunwind+mpi+papi+rocm+sampler+shared~sosflow amdgpu_target=gfx90a build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-sles15-zen3
[+]  pw6nc6k      ^adiak@0.2.2%gcc@11.2.0~ipo+mpi+shared build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-sles15-zen3
[+]  mj6xxft      ^cmake@3.26.3%gcc@11.2.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-sles15-zen3
[+]  igbrz2c          ^ncurses@6.3%gcc@11.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-sles15-zen3
[+]  hr2fmma          ^openssl@1.1.1-os%gcc@11.2.0~docs~shared build_system=generic certs=mozilla arch=linux-sles15-zen3
[+]  vw5wbnb      ^cray-mpich@8.1.23%gcc@11.2.0+wrappers build_system=generic arch=linux-sles15-zen3
[+]  mf2jtdz      ^elfutils@0.189%gcc@11.2.0+bzip2~debuginfod~nls+xz~zstd build_system=autotools arch=linux-sles15-zen3
[+]  gcwem7x          ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared build_system=generic arch=linux-sles15-zen3
[+]  yb3vc2g              ^diffutils@3.9%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  n325xld          ^m4@1.4.19%gcc@11.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-sles15-zen3
[+]  jok242c              ^libsigsegv@2.14%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  savxweu          ^pkgconf@1.8.0%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  yxqlxfz          ^xz@5.4.1%gcc@11.2.0+pic build_system=autotools libs=shared,static arch=linux-sles15-zen3
[+]  d35p7mx          ^zlib@1.2.13%gcc@11.2.0+optimize+pic+shared build_system=makefile arch=linux-sles15-zen3
[+]  6b72ibn      ^gmake@4.4.1%gcc@11.2.0~guile build_system=autotools arch=linux-sles15-zen3
[+]  vf4m3za      ^hip@5.3.0%gcc@11.2.0~cuda~ipo+rocm build_system=cmake build_type=Release generator=make patches=ca523f1 arch=linux-sles15-zen3
[+]  ssqrunh      ^hsa-rocr-dev@5.3.0%gcc@11.2.0+image~ipo+shared build_system=cmake build_type=Release generator=make patches=71e6851 arch=linux-sles15-zen3
[+]  s2mszhi      ^libunwind@1.6.2%gcc@11.2.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib build_system=autotools components=none libs=shared,static arch=linux-sles15-zen3
[+]  xikyk2u      ^llvm-amdgpu@5.3.0%gcc@11.2.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=a08bbe1 arch=linux-sles15-zen3
[+]  frkekrt      ^papi@6.0.0.1%gcc@11.2.0~cuda+example~infiniband~lmsensors~nvml~powercap~rapl~rocm~rocm_smi~sde+shared~static_tools build_system=autotools arch=linux-sles15-zen3
[+]  b4zevxj      ^python@3.7.15%gcc@11.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=linux-sles15-zen3
[+]  uo4mkvg          ^expat@2.5.0%gcc@11.2.0+libbsd build_system=autotools arch=linux-sles15-zen3
[+]  svce26u              ^libbsd@0.11.7%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  bcya2vp                  ^libmd@1.0.4%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  lggkkez          ^gdbm@1.23%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  ug2ciqw          ^gettext@0.21.1%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-sles15-zen3
[+]  hr2sqld              ^libiconv@1.17%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-sles15-zen3
[+]  op6kuq4              ^libxml2@2.10.3%gcc@11.2.0~python build_system=autotools arch=linux-sles15-zen3
[+]  uttw4m2              ^tar@1.34%gcc@11.2.0 build_system=autotools zip=pigz arch=linux-sles15-zen3
[+]  eqmk7ji                  ^pigz@2.7%gcc@11.2.0 build_system=makefile arch=linux-sles15-zen3
[+]  yiosjsu                  ^zstd@1.5.5%gcc@11.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-sles15-zen3
[+]  4zpefmz          ^libffi@3.4.4%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  xaieuhm          ^libxcrypt@4.4.33%gcc@11.2.0~obsolete_api build_system=autotools arch=linux-sles15-zen3
[+]  fr2owxn              ^perl@5.36.0%gcc@11.2.0+cpanm+open+shared+threads build_system=generic arch=linux-sles15-zen3
[+]  gqdvawb                  ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-sles15-zen3
[+]  rmgt2pq          ^readline@8.2%gcc@11.2.0 build_system=autotools patches=bbf97f1 arch=linux-sles15-zen3
[+]  eaxessg          ^sqlite@3.40.1%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-sles15-zen3
[+]  4qyr4mp          ^util-linux-uuid@2.38.1%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
</pre></details>

Error:
```
$> spack -e . install
...
==> Installing caliper-2.9.0-l4up667q34ctvbbo7vnqysrfpqmqivzo
==> No binary for caliper-2.9.0-l4up667q34ctvbbo7vnqysrfpqmqivzo found: installing from source
==> Using cached archive: /tmp/lpeyrala/PrgEnv-gnu/spack/var/spack/cache/_source-cache/archive/50/507ea74be64a2dfd111b292c24c4f55f459257528ba51a5242313fa50978371f.tar.gz
==> No patches needed for caliper
==> caliper: Executing phase: 'cmake'
==> caliper: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
...
4 errors found in build log:
     345    [ 45%] Building CXX object src/services/symbollookup/CMakeFiles/caliper-symbollookup.dir/LookupLibdw.cpp.o
     346    cd /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2/src/services/symbollookup && /opt/rocm-5.3.0/bin/hipcc  -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n
            2uymqrj6oj/spack-build-fkeuul2/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-bui
            ld-fkeuul2 -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/interface/c_fortran -I/tmp/
            lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2/src/services -I/tmp/lpeyrala/PrgEnv-gnu/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__s
            pack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/linux-sles15-zen3/gcc-11.2.0/elfutils-0.189-mf2jtdzxa4i4wr7divllf6kgi2iellfu/include -O2 -g -DNDEBUG -std=gn
            u++11 -fPIC -fvisibility=default -Wall -MD -MT src/services/symbollookup/CMakeFiles/caliper-symbollookup.dir/LookupLibdw.cpp.o -MF CMakeFiles/caliper-symbollookup.dir/LookupLibdw.cpp.o.d -o CMakeFiles/caliper-symbollookup.dir/LookupLibdw.cp
            p.o -c /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/services/symbollookup/LookupLibdw.cpp
     347    [ 46%] Building CXX object src/caliper/CMakeFiles/caliper-runtime.dir/builtin_configmanager.cpp.o
     348    cd /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2/src/caliper && /opt/rocm-5.3.0/bin/hipcc  -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spa
            ck-build-fkeuul2/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2 -I/
            tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/interface/c_fortran -I/opt/cray/pe/mpich/
            8.1.23/ofi/gnu/9.1/include -O2 -g -DNDEBUG -std=gnu++11 -fPIC -fvisibility=default -Wall -MD -MT src/caliper/CMakeFiles/caliper-runtime.dir/builtin_configmanager.cpp.o -MF CMakeFiles/caliper-runtime.dir/builtin_configmanager.cpp.o.d -o CMak
            eFiles/caliper-runtime.dir/builtin_configmanager.cpp.o -c /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/caliper/builtin_configmanager.cpp
     349    [ 47%] Building CXX object src/reader/CMakeFiles/caliper-reader.dir/Preprocessor.cpp.o
     350    cd /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2/src/reader && /opt/rocm-5.3.0/bin/hipcc  -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spac
            k-build-fkeuul2/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/include -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2 -I/t
            mp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src -I/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/interface/c_fortran -O2 -g -DNDEBUG -std=g
            nu++11 -fPIC -fvisibility=default -Wall -MD -MT src/reader/CMakeFiles/caliper-reader.dir/Preprocessor.cpp.o -MF CMakeFiles/caliper-reader.dir/Preprocessor.cpp.o.d -o CMakeFiles/caliper-reader.dir/Preprocessor.cpp.o -c /tmp/lpeyrala/spack-st
            age/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/reader/Preprocessor.cpp
  >> 351    /tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-src/src/services/sampler/Sampler.cpp:31:10: fatal error: 'libunwind.h' file not found
     352    #include <libunwind.h>
     353             ^~~~~~~~~~~~~
     354    1 error generated when compiling for gfx90a.
  >> 355    make[2]: *** [src/services/sampler/CMakeFiles/caliper-sampler.dir/build.make:79: src/services/sampler/CMakeFiles/caliper-sampler.dir/Sampler.cpp.o] Error 1
     356    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-caliper-2.9.0-fkeuul2kbsksrwwj2jmo6n2uymqrj6oj/spack-build-fkeuul2'
  >> 357    make[1]: *** [CMakeFiles/Makefile2:1389: src/services/sampler/CMakeFiles/caliper-sampler.dir/all] Error 2
...
```

### Error message

<details><summary>Error message</summary>
<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.20.0.dev0 (6ebc3f0b011a90752b602d4d3a2e215ff15c4a2c)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen3
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/11391126/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/11391127/spack-build-out.txt)

@daboehme @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/caliper/package.py
+++ b/var/spack/repos/builtin/packages/caliper/package.py
@@ -85,6 +85,7 @@ class Caliper(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+rocm+cuda")
 
     patch("for_aarch64.patch", when="target=aarch64:")
+    patch("sampler-service-missing-libunwind-include-dir.patch", when="@2.9.0 +rocm")
 
     def cmake_args(self):
         spec = self.spec

--- a/var/spack/repos/builtin/packages/caliper/package.py
+++ b/var/spack/repos/builtin/packages/caliper/package.py
@@ -85,7 +85,7 @@ class Caliper(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+rocm+cuda")
 
     patch("for_aarch64.patch", when="target=aarch64:")
-    patch("sampler-service-missing-libunwind-include-dir.patch", when="@2.9.0 +rocm")
+    patch("sampler-service-missing-libunwind-include-dir.patch", when="@2.9.0 +libunwind +sampler")
 
     def cmake_args(self):
         spec = self.spec

--- a/var/spack/repos/builtin/packages/caliper/sampler-service-missing-libunwind-include-dir.patch
+++ b/var/spack/repos/builtin/packages/caliper/sampler-service-missing-libunwind-include-dir.patch
@@ -1,0 +1,14 @@
+diff -ruN spack-src/src/services/sampler/CMakeLists.txt spack-src-patched/src/services/sampler/CMakeLists.txt
+--- spack-src/src/services/sampler/CMakeLists.txt	2022-11-30 13:52:42.000000000 -0500
++++ spack-src-patched/src/services/sampler/CMakeLists.txt	2023-05-04 20:43:47.240310306 -0400
+@@ -17,6 +17,10 @@
+ 
+ include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR})
+ 
++if (CALIPER_HAVE_LIBUNWIND)
++	  include_directories(${LIBUNWIND_INCLUDE_DIRS})
++endif()
++
+ add_library(caliper-sampler OBJECT ${CALIPER_SAMPLER_SOURCES})
+ 
+ add_service_objlib("caliper-sampler")

